IndieAuth support

gabek opened the pull request:
This is both the client and server-side support for IndieAuth.

https://indieauth.net/

- Adds new "chat settings" modal where IndieAuth is located.
- Authenticated users now display in chat with a checkmark.
- A display name in use by an authenticated user is now no longer available to other users.
- Adds IndieAuth server support so people can use Owncast itself as an IndieAuth server, primarily so people can authenticate at other Owncast servers, but also other IndieAuth sites as well.

gabek committed:
Able to authenticate user against IndieAuth. For #1273

gabek committed:
WIP server indieauth endpoint. For https://github.com/owncast/owncast/issues/1272

gabek committed:
Add migration to remove access tokens from user

gabek committed:
Add authenticated bool to user for display purposes

gabek committed:
Add indieauth modal and auth flair to display names. For #1273

gabek committed:
Validate URLs and display errors

gabek committed:
Renames, cleanups

gabek committed:
Handle relative auth endpoint paths. Add error handling for missing redirects.

gabek committed:
Disallow using display names in use by registered users. Closes #1810

gabek committed:
Verify code verifier via code challenge on callback

gabek committed:
Use relative path to authorization_endpoint

gabek committed:
Post-rebase fixes

aaronpk commented on auth/indieauth/client.go:
```diff
@@ -0,0 +1,87 @@
+	if err := json.Unmarshal(body, &response); err != nil {
+		return nil, nil, err
+	}
+
```
This needs some more error handling in case it gets back an error from the IndieAuth server before continuing on to assume there is a successful response returned.

gabek replied:
Since every implementation is unique, is it safe to rely on http status codes to determine there was an error as opposed to errors in the return payload? There's no way to know what the error structure will look like.

aaronpk replied:
The error structure is OAuth 2's structure with "error" and "error_description" keys in the JSON response

gabek replied:
I didn't realize that, I'm certainly not using that structure in my IndieAuth response. I'll make sure to do that, but if I didn't know it's possible other implementations didn't know either.

gabek replied:
What should be returned as `error` vs `error_description`?

aaronpk replied:
Yeah we chose to link out to the OAuth spec instead of copy all the error responses into the IndieAuth spec. Here's the list: https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

aaronpk commented on auth/indieauth/client.go:
```diff
@@ -0,0 +1,87 @@
+
+	// What we sent and what we got back must match
+	if meRequestVerifier != meResponseVerifier {
+		return nil, nil, errors.New("indieauth response does not match the initial anticipated auth destination")
```
Returning a JSON error here isn't great UX since this ends up being visible JSON to the user.

gabek replied:
A simple, but plain text, error is surfaced now. I filed #1838 to have something nicer.

aaronpk committed:
don't redirect unless a URL is present

avoids redirecting to `undefined` if there was an error

aaronpk committed:
improve error message if owncast server URL isn't set

aaronpk committed:
fix IndieAuth PKCE implementation

use SHA256 instead of SHA1, generates a longer code verifier (must be 43-128 chars long), fixes URL-safe SHA256 encoding

aaronpk commented on auth/indieauth/client.go:
```diff
@@ -0,0 +1,87 @@
+	data.Set("code_verifier", request.CodeVerifier)
+
+	codeChallenge := createCodeChallenge(request.CodeVerifier)
+	if codeChallenge != request.CodeChallenge {
```
What's the purpose of checking this here? The client stored the code verifier itself, and generated the code challenge in the previous request, so the code challenge here isn't coming from an external source, so it seems like the check is redundant.

gabek replied:
Removed redundant check.

aaronpk committed:
return real profile data for IndieAuth response

aaronpk commented on controllers/auth/indieauth/server.go:
```diff
@@ -0,0 +1,68 @@
+	// the IndieAuth flow.
+	u, err := url.Parse(redirectURI)
+	if err != nil {
+		controllers.WriteSimpleResponse(w, false, err.Error())
```
This needs to return a proper OAuth error response with `{"error":"...","error_description":"..."}`. Currently it sends back an Owncast-style JSON response

gabek replied:
hah! If I would have read your comment before today I would have known about the OAuth error response.

aaronpk committed:
check the code verifier in the IndieAuth server

aaronpk reviewed (commented):
I sent a PR in #1812 with some changes that I was able to do myself. The comments here are things I wasn't sure how you wanted to handle or didn't know enough Go to do myself.

aaronpk commented:
How would you feel about having this change the username to the `name` returned in the IndieAuth profile response after logging in? (Keeping the same uniqueness check of course). For owncast-to-owncast logins it would end up being the name of the server, and if they log in with their own website it would be the user's name. I'd imagine it would still send a username change event and everything so you'd see a notification like this in the chat:

> competent-newton is now known as Aaron Parecki

jamietanna commented:
Was there any reason we've hand-crafted the core part of the flow instead of using core go libraries for OAuth2?

E: I say this because recent changes to Indieauth allow us to use standard OAuth2 implementations now ☺ see https://gitlab.com/jamietanna/micropub-go/-/commit/76e6430bb90f6b74e418b5b311d4d3bdb3e472d8 as an example

jamietanna commented on auth/indieauth/helpers.go:
```diff
@@ -0,0 +1,120 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse html at remote auth host")
+	}
+	authorizationEndpointTag := cascadia.MustCompile("link[rel=authorization_endpoint]").MatchAll(scrapedHTMLDocument)
```
It'd be great if we could support the Discovery endpoint, as although fairly new it's going to be fairly widespread before long https://indieauth.spec.indieweb.org/#indieauth-server-metadata

jamietanna replied:
We could also outsource this logic to github.com/hawx/indieauth

gabek replied:
I apologize in advance for my ignorance, I'm quite detached from the ecosystem and how all these things work together. 
 But I'm confused. There's a not-yet-formalized update to IndieAuth, but a library that hasn't been updated in two years supports it?

I also didn't use an OAuth library because I wasn't building an OAuth implementation, I specifically was following the documentation around IndieAuth.  And not the "new" IndieAuth, the "original" IndieAuth (again I don't really understand the difference, or if they're compatible, etc).

Admittedly I'm a little frustrated by the entire thing and if IndieAuth is going to be an always changing spec that is not going to be used by people who visit Owncast streams I should probably cut my losses because I don't want the cost of maintaining something that's constantly changing but nobody is using.

jamietanna replied:
No I completely empathise with you and agree that it is a little frustrating!

Some of the recent changes we've made very recently have been to align much closer to OAuth2, so it's a much more straightforward implementation

And no you're right, that library doesn't yet support the new means, it was something I'm planning on contributing so we'd not need the logic in here either

gabek replied:
Filed an issue pointing to this thread to possibly work on the next version of IndieAuth if other servers start supporting it and people find the support within Owncast useful. https://github.com/owncast/owncast/issues/1837

jamietanna commented on auth/indieauth/helpers.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func createCodeChallenge(codeVerifier string) string {
+	hasher := sha1.New() // nolint:gosec
```
:thinking: does this create a SHA1 hash? PKCE is commonly used with SHA256

aaronpk replied:
I fixed this in my PR to this branch

jamietanna commented on auth/indieauth/response.go:
```diff
@@ -0,0 +1,16 @@
+// the IndieAuth flow.
+type Response struct {
+	Me      string  `json:"me"`
+	Profile Profile `json:"profile"`
```
this may not always be returned - what happens if it's absent?

gabek replied:
The profile isn't required. I might as well just remove it.

gabek replied:
Oh sorry, misread, you're looking at our server response.

There's currently no checks for these kind of things put in. That still needs to happen. IndieAuth support (client or server) will not be available unless you've set the required configuration (URL mostly). Errors need to be surfaced in the case that it's not configured.

jamietanna commented on auth/indieauth/client.go:
```diff
@@ -0,0 +1,87 @@
+		return nil, nil, errors.New("code challenge mismatch")
+	}
+
+	client := &http.Client{}
```
should we cache HTTP clients? Not sure if we do that elsewhere

gabek replied:
These requests will be rare, seems like that optimization wouldn't be needed. It's not something that's done elsewhere. Owncast doesn't do many outbound requests.

gabek commented:
All the feedback is completely appreciated, but this PR doesn't even pass tests and fails in every way. It's a draft PR that is looking less likely to be going in.

gabek committed:
Use a timestamp instead of a bool for authenticated

gabek committed:
Propertly handle and display error in modal

gabek committed:
Use auth'ed timestamp to derive authenticated flag to display in chat

gabek commented:
> How would you feel about having this change the username to the `name` returned in the IndieAuth profile response after logging in? (Keeping the same uniqueness check of course). For owncast-to-owncast logins it would end up being the name of the server, and if they log in with their own website it would be the user's name. I'd imagine it would still send a username change event and everything so you'd see a notification like this in the chat:
> 
> > competent-newton is now known as Aaron Parecki

This, and more, is something I did want to do. In fact my original vision was to set their display name, link their profile, and even put a little Owncast logo if what they're authenticating against is an Owncast server. But then I realized I'm making a lot of assumptions about what data they want to make public. While they may want to authenticate against a service, that doesn't mean they want their identity, as known by that service, to be made public. Even sharing that this person has an Owncast instance is making an assumption that they want others to know about it.

I could, of course, add more switches, knobs and prompts to ask people what they want to share, but I think that would just add confusion and make the whole flow feel like Facebook Connect, where Facebook would hand over data about a user when you used them for authentication purposes. So saying it's auth, and nothing more, is probably the most fair way to go about it.

aaronpk commented:
That's a fair point. The way this usually works with IndieAuth servers is the server will show a prompt to indicate what info will be shared with the site you're logging in to, just like you see in other OAuth servers. There's also scopes that the site can request to tell the server whether it wants to know their name or email in addition to just doing the authentication.

gabek committed:
Linting

gabek committed:
Add new chat settings modal anad split up indieauth ui

gabek committed:
Remove logging error

gabek committed:
Update the IndieAuth modal UI. For #1273

gabek committed:
Add IndieAuth repsonse error checking

gabek committed:
Disable IndieAuth client if server URL is not set.

gabek committed:
Add explicit error messages for specific error types

gabek committed:
Fix bad logic

gabek committed:
Return OAuth-keyed error responses for indieauth server

gabek committed:
Display IndieAuth error in plain text with link to return to main page

gabek committed:
Remove redundant check

gabek committed:
Add additional detail to error

gabek committed:
Hide IndieAuth details behind disclosure details

gabek commented:
I think it's in better shape now. Hopefully the above feedback is implemented as intended.  I'm guessing there's still some rough edges in real world use, however.

gabek committed:
Break out migration into two steps because some people have been runing dev in production

gabek committed:
Add auth option to user dropdown

gabek commented:
I'm going to merge this into `develop` so daily development can live side by side with this and issues can be surfaced. If it all falls apart it's easy enough to turn off.
